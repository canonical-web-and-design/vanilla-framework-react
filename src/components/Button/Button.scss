@import 'vanilla-framework/scss/base_button';
@import 'vanilla-framework/scss/patterns_buttons';
@include vf-p-buttons;

body {
  background: #cdcdcd;
}