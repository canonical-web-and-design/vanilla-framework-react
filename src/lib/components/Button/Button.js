import React from 'react';
import PropTypes from 'prop-types';
import './Button.scss';

class Button extends React.Component {
  constructor() {
    super();
    this.onClick = this.onClick.bind(this);
    this.getClassString = this.getClassString.bind(this);
  }

  onClick(event) {
    if (!this.props.disabled) {
      this.props.onClick(event);
    }
  }

  getClassString() {
    let classString = 'p-button--base';

    if (this.props.positive) classString = 'p-button--positive';
    else if (this.props.negative) classString = 'p-button--negative';
    else if (this.props.brand) classString = 'p-button--brand';
    else if (this.props.neutral) classString = 'p-button--neutral';
    if (this.props.inline) classString = `${classString} is-inline`;

    return classString;
  }

  render() {
    return (
      <button
        className={this.getClassString()}
        onClick={this.props.onClick}
        disabled={this.props.disabled}
      >
        {this.props.value}
      </button>
    );
  }
}

Button.defaultProps = {
  neutral: false,
  brand: false,
  negative: false,
  positive: false,
  disabled: false,
  inline: false,
  onClick: () => null,
};

Button.propTypes = {
  value: PropTypes.string.isRequired,
  neutral: PropTypes.bool,
  brand: PropTypes.bool,
  negative: PropTypes.bool,
  positive: PropTypes.bool,
  inline: PropTypes.bool,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.displayName = 'Button';

export default Button;
