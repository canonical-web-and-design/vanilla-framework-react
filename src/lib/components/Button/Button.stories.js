import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import Button from './Button';

storiesOf('Buttons', module)
  .add('Base',
    withInfo('A base button is usually used alongside a neutral button. Precedence for button type is: base < neutral < brand < negative < positive.')(() =>
      <Button value="Base button 😶" />,
    ),
  )
  .add('Neutral',
    withInfo('A neutral button can be used to indicate a positive action that isn\'t necessarily the main call-to-action.')(() =>
      <Button neutral value="Neutral button 😐" />,
    ),
  )
  .add('Brand',
    withInfo('You can use the brand button with the main color of your brand.')(() =>
      <Button brand value="Brand button 💥" />,
    ),
  )
  .add('Negative',
    withInfo('A negative button can be used to indicate a negative action that is destructive or permanent.')(() =>
      <Button negative value="Negative button 😡" />,
    ),
  )
  .add('Positive',
    withInfo('A positive button can be used to indicate a positive action that is the main call-to-action.')(() =>
      <Button positive value="Positive button 😁" />,
    ),
  )
  .add('Inline',
    withInfo('Should you wish to place a button after a line of inline text, as a CTA for example, you can do so by adding the prop "inline" to the Button component.')(() => (
      <p>
        This is an inline button. <Button neutral inline value="Inline button" />
      </p>),
    ),
  )
  .add('Disabled',
    withInfo('You can disable a button by adding the "disabled" prop.')(() =>
      <Button disabled value="Disabled button" />,
    ),
  );
