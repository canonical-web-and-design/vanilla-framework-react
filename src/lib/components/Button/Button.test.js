import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import Button from './Button';

describe('<Button>', () => {
  it('should render a simple Button correctly', () => {
    const button = ReactTestRenderer.create(
      <Button value="Submit" />);
    const json = button.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a modified Button correctly', () => {
    const button = ReactTestRenderer.create(
      <div>
        <Button neutral value="Neutral" />
        <Button negative value="Negative" />
        <Button positive value="Positive" />
        <Button brand value="Brand" />
        <Button inline value="Inline" />
        <Button disabled value="Disabled" />
      </div>);
    const json = button.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('should render a button element by default', () => {
    const button = shallow(<Button>Button</Button>);
    expect(button.find('button').hostNodes().length).toBe(1);
    expect(button.find('a').hostNodes().length).toBe(0);
  });

  it('should render an anchor element if href exists', () => {
    const button = shallow(<Button href="/link">Link</Button>);
    expect(button.find('a').hostNodes().length).toBe(1);
    expect(button.find('button').hostNodes().length).toBe(0);
  });

  it('should be the same whether children or value prop is passed through', () => {
    const text = 'Value';
    const buttonChild = shallow(<Button>{text}</Button>);
    const buttonValue = shallow(<Button value={text} />);
    expect(buttonChild.html()).toBe(buttonValue.html());

    const newButtonValue = shallow(<Button value="New Value" />);
    expect(buttonChild.html()).not.toBe(newButtonValue.html());
  });

  it('calls onClick if it exists', () => {
    const onClick = jest.fn();
    const button = mount(<Button onClick={onClick}>onClick test</Button>);

    button.find('button').hostNodes().simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  it('does not call onClick if it exists and is disabled', () => {
    const onClick = jest.fn();
    const button = mount(<Button disabled onClick={onClick}>onClick test</Button>);

    button.find('button').hostNodes().simulate('click');
    expect(onClick).not.toHaveBeenCalled();
  });
});

