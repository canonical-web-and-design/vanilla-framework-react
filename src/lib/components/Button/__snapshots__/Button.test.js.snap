// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render a modified Button correctly 1`] = `
<div>
  <button
    className="p-button--neutral"
    disabled={false}
    href={undefined}
    onClick={[Function]}
  >
    Neutral
  </button>
  <button
    className="p-button--negative"
    disabled={false}
    href={undefined}
    onClick={[Function]}
  >
    Negative
  </button>
  <button
    className="p-button--positive"
    disabled={false}
    href={undefined}
    onClick={[Function]}
  >
    Positive
  </button>
  <button
    className="p-button--brand"
    disabled={false}
    href={undefined}
    onClick={[Function]}
  >
    Brand
  </button>
  <button
    className="p-button--base is-inline"
    disabled={false}
    href={undefined}
    onClick={[Function]}
  >
    Inline
  </button>
  <button
    className="p-button--base is--disabled"
    disabled={true}
    href={undefined}
    onClick={[Function]}
  >
    Disabled
  </button>
</div>
`;

exports[`<Button> should render a simple Button correctly 1`] = `
<button
  className="p-button--base"
  disabled={false}
  href={undefined}
  onClick={[Function]}
>
  Submit
</button>
`;
